whisper/whisperv5/whisper.go

```diff
@@ -436,22 +453,55 @@ func (wh *Whisper) add(envelope *Envelope) error {
 	wh.poolMu.Unlock()
 
 	if alreadyCached {
-		glog.V(logger.Detail).Infof("whisper envelope already cached: %x\n", envelope)
+		glog.V(logger.Detail).Infof("whisper envelope already cached [%x]\n", envelope.Hash())
 	} else {
+		glog.V(logger.Detail).Infof("cached whisper envelope [%x]: %v\n", envelope.Hash(), envelope)
 		wh.postEvent(envelope, messagesCode) // notify the local node about the new message
-		glog.V(logger.Detail).Infof("cached whisper envelope %v\n", envelope)
 	}
 	return nil
 }
 
-// postEvent delivers the message to the watchers.
+// postEvent queues the message for further processing.
 func (w *Whisper) postEvent(envelope *Envelope, messageCode uint64) {
 	// if the version of incoming message is higher than
 	// currently supported version, we can not decrypt it,
 	// and therefore just ignore this message
 	if envelope.Ver() <= EnvelopeVersion {
-		// todo: review if you need an additional thread here
-		go w.filters.NotifyWatchers(envelope, messageCode)
+		w.checkOverflow()
+		i := QueueItem{hash: envelope.Hash(), code: messageCode}
+		w.messageQueue <- i
+	}
+}
+
+// checkOverflow checks if message queue overflow occurs and reports it if necessary.
+func (w *Whisper) checkOverflow() {
+	queueSize := len(w.messageQueue)
+
+	if queueSize == messageQueueLimit {
+		if !w.overflow {
+			w.overflow = true
+			glog.V(logger.Warn).Infoln("message queue overflow")
+		}
+	} else if queueSize <= messageQueueLimit/2 {
+		if w.overflow {
+			w.overflow = false
+		}
+	}
+}
+
+// processQueue delivers the messages to the watchers during the lifetime of the whisper node.
+func (w *Whisper) processQueue() {
+	for {
+		select {
+		case <-w.quit:
+			return
+
+		case i := <-w.messageQueue:
+			w.poolMu.Lock()
```

Comment:
as for the messageCode i think it is not meant to be here. 
whether it is from a trusted peer or not should be the attribute of the envelop set by the messagehandler
also you could just use the pool to check duplicates only when you first get the envelope 
then you would check expiry next when processqueue worker receives it or when you call Envelopes
and check it also when messages are retrieved via filter or Messages
That way you could get rid of the update loop doing expiries 

AFAIU currently Envelopes and maybe Messages too can actually return expired items no?